[Error] Crash In V7.0.0

### Confirmation

- [x] I have read the [Code of Conduct](https://github.com/Hamza417/Peristyle/blob/master/CODE_OF_CONDUCT.md)
- [x] I have searched the existing issues

### Error Description

The entire new Live Auto  Wallpaper feature is not working for me and all I get is a black wallpaper when I tap "Live Auto Wallpaper".

### Steps to Reproduce

1. Tap the Live Wallpapers button in the home screen (the one next to the Folders button).
2. Select Live Auto Wallpaper.
3. Tap Live Auto Wallpaper.
4. Tap Set wallpaper in the Live Wallpaper Picker android app.
5. Select Home screen.
6. Return to Peristyle.

### Expected Behavior

No crash lol.

### Error Logs

java.lang.RuntimeException: Unable to start service app.simple.peri.services.LiveAutoWallpaperService@cffa972 with Intent { act=action.PREVIEW_WALLPAPER cmp=app.simple.peri/.services.LiveAutoWallpaperService (has extras) }: java.lang.NullPointerException: Attempt to invoke virtual method 'boolean java.lang.Class.isInterface()' on a null object reference
	at android.app.ActivityThread.handleServiceArgs(ActivityThread.java:4864)
	at android.app.ActivityThread.-$$Nest$mhandleServiceArgs(Unknown Source:0)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2254)
	at android.os.Handler.dispatchMessage(Handler.java:106)
	at android.os.Looper.loopOnce(Looper.java:210)
	at android.os.Looper.loop(Looper.java:299)
	at android.app.ActivityThread.main(ActivityThread.java:8247)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:559)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:954)
Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'boolean java.lang.Class.isInterface()' on a null object reference
	at java.lang.Class.isAssignableFrom(Class.java:890)
	at android.os.Parcel.readParcelableCreatorInternal(Parcel.java:4853)
	at android.os.Parcel.readParcelableInternal(Parcel.java:4766)
	at android.os.Parcel.readValue(Parcel.java:4532)
	at android.os.Parcel.readValue(Parcel.java:4312)
	at android.os.Parcel.-$$Nest$mreadValue(Unknown Source:0)
	at android.os.Parcel$LazyValue.apply(Parcel.java:4410)
	at android.os.Parcel$LazyValue.apply(Parcel.java:4369)
	at android.os.BaseBundle.getValueAt(BaseBundle.java:394)
	at android.os.BaseBundle.getValue(BaseBundle.java:374)
	at android.os.BaseBundle.getValue(BaseBundle.java:357)
	at android.os.BaseBundle.get(BaseBundle.java:696)
	at android.os.Bundle.getParcelable(Bundle.java:947)
	at android.content.Intent.getParcelableExtra(Intent.java:8927)
	at androidx.activity.ComponentDialog$$ExternalSyntheticApiModelOutline0.m$2(SourceFile:1)
	at app.simple.peri.services.LiveAutoWallpaperService.onStartCommand(SourceFile:51)
	at android.app.ActivityThread.handleServiceArgs(ActivityThread.java:4846)
	... 9 more

### Device Information

Device: POCO F5 PRO ; OS Android 13.

### Additional Context

It seems like the Wallpaper Manager API options does nothing after this crash. Tapping Next Wallpaper does nothing. However, I can still change the wallpaper using the Set as Wallpaper button.

### Sufficiency Confirmation

- [x] I have provided all the information requested above and deemed sufficient for the developer to address the issue

There is an improvement but it's still buggy.

When I tap the circular Live Wallpapers button then tap the Live Auto Wallpaper button one of the following can happen:
A. The app crashes and the wallpaper gets set to the device's default wallpaper.
B. The Live Wallpaper Picker is opened but after tapping Set wallpaper, then Home screen, a black wallpaper is set. Also, the Live Wallpaper Picker remains in the recent apps and when I reopen it the app crashes.
C. If after the Live Wallpaper Picker is opened, you go back and swipe the app from the recent apps and tap the Live Auto Wallpaper button again the app crashes.

The good news is that after setting a wallpaper manually, the automatic wallpaper changer starts working again. Also, I'm not sure but I think like a few minutes after crash C the wallpaper starts auto changing again.

```
java.lang.RuntimeException: Unable to start service app.simple.peri.services.LiveAutoWallpaperService@9ea5a3 with Intent { act=action.PREVIEW_WALLPAPER cmp=app.simple.peri/.services.LiveAutoWallpaperService (has extras) }: java.lang.NullPointerException: Attempt to invoke virtual method 'boolean java.lang.Class.isInterface()' on a null object reference
	at android.app.ActivityThread.handleServiceArgs(ActivityThread.java:4864)
	at android.app.ActivityThread.-$$Nest$mhandleServiceArgs(Unknown Source:0)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2254)
	at android.os.Handler.dispatchMessage(Handler.java:106)
	at android.os.Looper.loopOnce(Looper.java:210)
	at android.os.Looper.loop(Looper.java:299)
	at android.app.ActivityThread.main(ActivityThread.java:8247)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:559)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:954)
Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'boolean java.lang.Class.isInterface()' on a null object reference
	at java.lang.Class.isAssignableFrom(Class.java:890)
	at android.os.Parcel.readParcelableCreatorInternal(Parcel.java:4853)
	at android.os.Parcel.readParcelableInternal(Parcel.java:4766)
	at android.os.Parcel.readValue(Parcel.java:4532)
	at android.os.Parcel.readValue(Parcel.java:4312)
	at android.os.Parcel.-$$Nest$mreadValue(Unknown Source:0)
	at android.os.Parcel$LazyValue.apply(Parcel.java:4410)
	at android.os.Parcel$LazyValue.apply(Parcel.java:4369)
	at android.os.BaseBundle.getValueAt(BaseBundle.java:394)
	at android.os.BaseBundle.getValue(BaseBundle.java:374)
	at android.os.BaseBundle.getValue(BaseBundle.java:357)
	at android.os.BaseBundle.get(BaseBundle.java:696)
	at android.os.Bundle.getParcelable(Bundle.java:947)
	at android.content.Intent.getParcelableExtra(Intent.java:8927)
	at androidx.activity.ComponentDialog$$ExternalSyntheticApiModelOutline0.m$2(SourceFile:1)
	at app.simple.peri.services.LiveAutoWallpaperService.onStartCommand(SourceFile:51)
	at android.app.ActivityThread.handleServiceArgs(ActivityThread.java:4846)
	... 9 more
```

nvm managed to replicate it

java.lang.RuntimeException: Unable to start service app.simple.peri.services.LiveAutoWallpaperService@ac4d2f7 with Intent { act=action.PREVIEW_WALLPAPER cmp=app.simple.peri/.services.LiveAutoWallpaperService (has extras) }: java.lang.NullPointerException: Attempt to invoke virtual method 'boolean java.lang.Class.isInterface()' on a null object reference
	at android.app.ActivityThread.handleServiceArgs(ActivityThread.java:5261)
	at android.app.ActivityThread.-$$Nest$mhandleServiceArgs(Unknown Source:0)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2447)
	at android.os.Handler.dispatchMessage(Handler.java:106)
	at android.os.Looper.loopOnce(Looper.java:226)
	at android.os.Looper.loop(Looper.java:313)
	at android.app.ActivityThread.main(ActivityThread.java:8762)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:604)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1067)
Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'boolean java.lang.Class.isInterface()' on a null object reference
	at java.lang.Class.isAssignableFrom(Class.java:890)
	at android.os.Parcel.readParcelableCreatorInternal(Parcel.java:4865)
	at android.os.Parcel.readParcelableInternal(Parcel.java:4778)
	at android.os.Parcel.readValue(Parcel.java:4544)
	at android.os.Parcel.readValue(Parcel.java:4324)
	at android.os.Parcel.-$$Nest$mreadValue(Unknown Source:0)
	at android.os.Parcel$LazyValue.apply(Parcel.java:4422)
	at android.os.Parcel$LazyValue.apply(Parcel.java:4381)
	at android.os.BaseBundle.getValueAt(BaseBundle.java:394)
	at android.os.BaseBundle.getValue(BaseBundle.java:374)
	at android.os.BaseBundle.getValue(BaseBundle.java:357)
	at android.os.BaseBundle.get(BaseBundle.java:696)
	at android.os.Bundle.getParcelable(Bundle.java:947)
	at android.content.Intent.getParcelableExtra(Intent.java:9558)
	at androidx.activity.ComponentDialog$$ExternalSyntheticApiModelOutline0.m$2(SourceFile:1)
	at app.simple.peri.services.LiveAutoWallpaperService.onStartCommand(SourceFile:51)
	at android.app.ActivityThread.handleServiceArgs(ActivityThread.java:5243)
	... 9 more

--- a/app/build.gradle
+++ b/app/build.gradle
@@ -18,8 +18,8 @@ android {
         applicationId "app.simple.peri"
         minSdk 24
         targetSdk 35
-        versionCode 704
-        versionName "v7.0.4-rc4"
+        versionCode 701
+        versionName "v7.0.1-rc5"
 
         testInstrumentationRunner "androidx.test.runner.AndroidJUnitRunner"
 

--- a/app/build.gradle
+++ b/app/build.gradle
@@ -18,8 +18,8 @@ android {
         applicationId "app.simple.peri"
         minSdk 24
         targetSdk 35
-        versionCode 701
-        versionName "v7.0.1"
+        versionCode 703
+        versionName "v7.0.3-rc3"
 
         testInstrumentationRunner "androidx.test.runner.AndroidJUnitRunner"
 

--- a/app/build.gradle
+++ b/app/build.gradle
@@ -18,8 +18,8 @@ android {
         applicationId "app.simple.peri"
         minSdk 24
         targetSdk 35
-        versionCode 703
-        versionName "v7.0.3-rc3"
+        versionCode 704
+        versionName "v7.0.4-rc4"
 
         testInstrumentationRunner "androidx.test.runner.AndroidJUnitRunner"
 

--- a/app/src/main/java/app/simple/peri/abstraction/AbstractAutoLiveWallpaperService.kt
+++ b/app/src/main/java/app/simple/peri/abstraction/AbstractAutoLiveWallpaperService.kt
@@ -54,6 +54,7 @@ abstract class AbstractAutoLiveWallpaperService : AbstractComposeAutoWallpaperSe
             val intent = LiveAutoWallpaperService.getIntent(applicationContext, LiveAutoWallpaperService.NEXT_WALLPAPER)
             intent.putExtra(LiveAutoWallpaperService.EXTRA_WALLPAPER, wallpaper as Parcelable)
             applicationContext.startService(intent)
+            Log.i(TAG, "wallpaper data sent, live wallpaper should handle this from here")
         } else {
             super.setSameWallpaper(wallpaper)
         }
@@ -64,6 +65,7 @@ abstract class AbstractAutoLiveWallpaperService : AbstractComposeAutoWallpaperSe
             val intent = LiveAutoWallpaperService.getIntent(applicationContext, LiveAutoWallpaperService.NEXT_WALLPAPER)
             intent.putExtra(LiveAutoWallpaperService.EXTRA_WALLPAPER, wallpaper as Parcelable)
             applicationContext.startService(intent)
+            Log.i(TAG, "wallpaper data sent, live wallpaper should handle this from here")
         } else {
             super.setHomeScreenWallpaper(wallpaper)
         }

--- a/app/src/main/java/app/simple/peri/abstraction/AutoWallpaperUtils.kt
+++ b/app/src/main/java/app/simple/peri/abstraction/AutoWallpaperUtils.kt
@@ -5,12 +5,13 @@ import android.graphics.BitmapFactory
 import android.graphics.Rect
 import android.os.Build
 import android.util.Log
-import app.simple.peri.abstraction.AbstractComposeAutoWallpaperService.Companion.TAG
 import app.simple.peri.utils.BitmapUtils
 import app.simple.peri.utils.FileUtils.toFile
 import java.io.ByteArrayInputStream
 
 object AutoWallpaperUtils {
+
+    private const val TAG = "AutoWallpaperUtils"
 
     fun getBitmapFromFile(path: String, expectedWidth: Int, expectedHeight: Int, crop: Boolean = true, recycle: Boolean = true, onBitmap: (Bitmap) -> Unit) {
         path.toFile().inputStream().use { stream ->

--- a/app/src/main/java/app/simple/peri/compose/screens/Wallpaper.kt
+++ b/app/src/main/java/app/simple/peri/compose/screens/Wallpaper.kt
@@ -67,6 +67,7 @@ import app.simple.peri.constants.Misc
 import app.simple.peri.factories.TagsViewModelFactory
 import app.simple.peri.models.Effect
 import app.simple.peri.models.Wallpaper
+import app.simple.peri.services.LiveAutoWallpaperService
 import app.simple.peri.utils.BitmapUtils.applyEffects
 import app.simple.peri.utils.BitmapUtils.multiplyMatrices
 import app.simple.peri.utils.ConditionUtils.isNotNull
@@ -351,6 +352,12 @@ fun Wallpaper(navController: NavHostController, associatedWallpaper: Wallpaper? 
                                 drawable = BitmapDrawable(context.resources, bitmap)
                                 showScreenSelectionDialog = true
                                 showWallpaperLaunchedEffect.value = false
+
+                                try {
+                                    context.stopService(LiveAutoWallpaperService.getIntent(context))
+                                } catch (e: IllegalStateException) {
+                                    Log.e("Wallpaper", "Service not running")
+                                }
                             }
                         }
                     }

--- a/app/src/main/java/app/simple/peri/services/AutoWallpaperService.kt
+++ b/app/src/main/java/app/simple/peri/services/AutoWallpaperService.kt
@@ -78,7 +78,7 @@ class AutoWallpaperService : AbstractAutoLiveWallpaperService() {
 
     override fun onDestroy() {
         super.onDestroy()
-        Log.d(TAG, "Service destroyed")
+        Log.d(TAG, "auto wallpaper service destroyed")
     }
 
     private fun init() {

--- a/app/src/main/java/app/simple/peri/services/LiveAutoWallpaperService.kt
+++ b/app/src/main/java/app/simple/peri/services/LiveAutoWallpaperService.kt
@@ -43,8 +43,12 @@ class LiveAutoWallpaperService : WallpaperService() {
         handler = Handler(Looper.getMainLooper()) { msg ->
             when (msg.what) {
                 MSG_SET_WALLPAPER -> {
-                    val filePath = msg.obj as String
-                    engine!!.setWallpaper(filePath)
+                    try {
+                        val filePath = msg.obj as String
+                        engine!!.setWallpaper(filePath)
+                    } catch (e: NullPointerException) {
+                        e.printStackTrace()
+                    }
                 }
             }
             true

--- a/app/src/main/java/app/simple/peri/services/LiveAutoWallpaperService.kt
+++ b/app/src/main/java/app/simple/peri/services/LiveAutoWallpaperService.kt
@@ -18,6 +18,7 @@ import app.simple.peri.models.Wallpaper
 import app.simple.peri.preferences.MainComposePreferences
 import app.simple.peri.preferences.SharedPreferences
 import app.simple.peri.utils.BitmapUtils.applyEffects
+import app.simple.peri.utils.ConditionUtils.invert
 import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.launch
@@ -113,26 +114,30 @@ class LiveAutoWallpaperService : WallpaperService() {
 
         fun setWallpaper(filePath: String) {
             CoroutineScope(Dispatchers.Main).launch {
-                val canvas = surfaceHolder.lockCanvas()
-                var localBitmap: Bitmap? = null
+                if (surfaceHolder.surface.isValid) {
+                    val canvas = surfaceHolder.lockCanvas()
+                    var localBitmap: Bitmap? = null
 
-                withContext(Dispatchers.Default) {
-                    try {
-                        getBitmapFromFile(filePath, canvas.width, canvas.height, recycle = false) { bmp ->
-                            localBitmap = bmp
-                        }
-                    } catch (e: NullPointerException) {
-                        if (localBitmap == null) {
-                            localBitmap = bitmap
+                    withContext(Dispatchers.Default) {
+                        try {
+                            getBitmapFromFile(filePath, canvas.width, canvas.height, recycle = false) { bmp ->
+                                localBitmap = bmp
+                            }
+                        } catch (e: NullPointerException) {
+                            if (localBitmap == null) {
+                                localBitmap = bitmap
+                            }
                         }
                     }
-                }
 
-                if (localBitmap != null) {
-                    if (canvas != null) {
-                        surfaceHolder.unlockCanvasAndPost(canvas)
-                        setBitmapWithCrossfade(localBitmap!!)
+                    if (localBitmap != null) {
+                        if (canvas != null) {
+                            surfaceHolder.unlockCanvasAndPost(canvas)
+                            setBitmapWithCrossfade(localBitmap!!)
+                        }
                     }
+                } else {
+                    Log.i(TAG, "Surface is not valid, skipping wallpaper change")
                 }
             }
         }
@@ -214,11 +219,16 @@ class LiveAutoWallpaperService : WallpaperService() {
         }
 
         override fun onSurfaceDestroyed(holder: SurfaceHolder?) {
-            super.onSurfaceDestroyed(holder)
-            bitmap?.recycle()
-            bitmap = null
-            oldBitmap?.recycle()
-            oldBitmap = null
+            if (isPreview.invert()) {
+                super.onSurfaceDestroyed(holder)
+                Log.i(TAG, "wallpaper surface destroyed")
+                bitmap?.recycle()
+                bitmap = null
+                oldBitmap?.recycle()
+                oldBitmap = null
+            } else {
+                Log.i(TAG, "Preview mode, skipping destruction")
+            }
         }
     }
 

--- a/app/src/main/java/app/simple/peri/services/LiveAutoWallpaperService.kt
+++ b/app/src/main/java/app/simple/peri/services/LiveAutoWallpaperService.kt
@@ -87,7 +87,7 @@ class LiveAutoWallpaperService : WallpaperService() {
         when (intent?.action) {
             NEXT_WALLPAPER, PREVIEW_WALLPAPER -> {
                 try {
-                    val wallpaper: Wallpaper = IntentCompat.getParcelableExtra(intent, EXTRA_WALLPAPER, Wallpaper::class.java)!!
+                    val wallpaper = IntentCompat.getParcelableExtra(intent, EXTRA_WALLPAPER, Wallpaper::class.java)!!
                     Log.i(TAG, "Setting wallpaper: ${wallpaper.filePath}")
                     val msg = handler?.obtainMessage(MSG_SET_WALLPAPER, wallpaper.filePath)!!
                     handler?.sendMessage(msg)
@@ -129,8 +129,10 @@ class LiveAutoWallpaperService : WallpaperService() {
                 }
 
                 if (localBitmap != null) {
-                    surfaceHolder.unlockCanvasAndPost(canvas)
-                    setBitmapWithCrossfade(localBitmap!!)
+                    if (canvas != null) {
+                        surfaceHolder.unlockCanvasAndPost(canvas)
+                        setBitmapWithCrossfade(localBitmap!!)
+                    }
                 }
             }
         }
@@ -205,7 +207,9 @@ class LiveAutoWallpaperService : WallpaperService() {
                     }
                 }
 
-                it.unlockCanvasAndPost(canvas)
+                if (canvas != null) {
+                    it.unlockCanvasAndPost(canvas)
+                }
             }
         }
 

--- a/app/src/main/java/app/simple/peri/services/LiveAutoWallpaperService.kt
+++ b/app/src/main/java/app/simple/peri/services/LiveAutoWallpaperService.kt
@@ -12,12 +12,12 @@ import android.os.Looper
 import android.service.wallpaper.WallpaperService
 import android.util.Log
 import android.view.SurfaceHolder
+import androidx.core.content.IntentCompat
 import app.simple.peri.abstraction.AutoWallpaperUtils.getBitmapFromFile
 import app.simple.peri.models.Wallpaper
 import app.simple.peri.preferences.MainComposePreferences
 import app.simple.peri.preferences.SharedPreferences
 import app.simple.peri.utils.BitmapUtils.applyEffects
-import app.simple.peri.utils.ParcelUtils.parcelable
 import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.launch
@@ -87,7 +87,7 @@ class LiveAutoWallpaperService : WallpaperService() {
         when (intent?.action) {
             NEXT_WALLPAPER, PREVIEW_WALLPAPER -> {
                 try {
-                    val wallpaper: Wallpaper = intent.parcelable<Wallpaper>(EXTRA_WALLPAPER)!!
+                    val wallpaper: Wallpaper = IntentCompat.getParcelableExtra(intent, EXTRA_WALLPAPER, Wallpaper::class.java)!!
                     Log.i(TAG, "Setting wallpaper: ${wallpaper.filePath}")
                     val msg = handler?.obtainMessage(MSG_SET_WALLPAPER, wallpaper.filePath)!!
                     handler?.sendMessage(msg)

--- a/app/src/main/java/app/simple/peri/services/LiveAutoWallpaperService.kt
+++ b/app/src/main/java/app/simple/peri/services/LiveAutoWallpaperService.kt
@@ -257,5 +257,9 @@ class LiveAutoWallpaperService : WallpaperService() {
                 this.action = action
             }
         }
+
+        fun getIntent(context: Context): Intent {
+            return Intent(context, LiveAutoWallpaperService::class.java)
+        }
     }
 }

